### PR TITLE
Scope distribution validation disabling to Dynamo tracing

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185438

TorchDynamo permanently disabled torch.distributions parameter validation from
TorchPatcher.patch() when a user constructed a torch.compile wrapper. That
changed Distribution._validate_args for the whole Python process, so unrelated
eager distribution constructors could silently accept invalid parameters after
any torch.compile use.

Move the validation suppression out of the one-time patcher and into scoped
Dynamo capture/execution regions. The normal OptimizeContext path now disables
validation only while the optimized wrapper is active and restores the previous
user default afterward. AOT fullgraph capture uses the same scoped suppression
around fullgraph_capture so direct aot_compile_fullgraph, aot_compile_module,
and torch.compile(...).aot_compile keep the existing ability to trace
Distributions code whose validation path enters skipped constraints helpers.

The scoped helper is process-global because Distribution._validate_args is
process-global. It uses a lock and depth counter so overlapping scopes do not
restore validation while another Dynamo scope is still active. Cleanup callbacks
are drained before re-raising cleanup failures, and the fullgraph no-compiled-
frames error is deferred until after cleanup, so exceptional paths do not leak
validation=False globally.

A simpler deletion of the old set_default_validate_args(False) call was tested,
but it breaks existing distribution tracing because validation enters
constraints.is_dependent, which Dynamo intentionally skips. Keeping the disable
scoped preserves that behavior without mutating user-visible global state.

Fixes #157926
Generated by my agent

Benchmark Results:
Microbenchmark command measured a cached torch.compile(..., backend="eager")
call for a one-op function over 20,000 calls, repeated 7 times. The baseline
approximates the previous behavior by disabling validation globally once and
using a no-op per-call distribution hook.

baseline_global_disabled_no_scope_hook raw us/call:
13.942, 14.266, 14.079, 14.850, 14.161, 14.121, 14.248
median: 14.161 us/call

patched_scoped_validation_hook raw us/call:
16.378, 15.695, 15.391, 15.386, 15.434, 15.515, 15.265
median: 15.434 us/call

Test Plan:
python -m py_compile torch/_dynamo/aot_compile.py torch/_dynamo/eval_frame.py test/dynamo/test_aot_compile.py test/dynamo/test_misc.py
python test/dynamo/test_misc.py MiscTests.test_distribution_validation_restored_on_fullgraph_no_compile_error MiscTests.test_distribution_validation_restored_if_earlier_cleanup_raises MiscTests.test_distribution_validation_scope_allows_non_lifo_cleanup MiscTests.test_torch_compile_preserves_distribution_validation_default MiscTests.test_torch_distributions_gamma_dynamic MiscTests.test_torch_distributions_lazy_property
python test/dynamo/test_aot_compile.py TestAOTCompile.test_aot_compile_scopes_distribution_validation TestAOTCompile.test_aot_compile_fullgraph_scopes_distribution_validation TestAOTCompile.test_aot_compile_fullgraph_with_existing_guard_filter_fn TestAOTCompile.test_aot_compile_module_scopes_distribution_validation
python test/dynamo/test_functions.py FunctionTests.test_torch_distributions_functions
python test/dynamo/test_modes.py TorchFunctionModeTests.test_expand
python test/dynamo/test_repros.py ReproTests.test_distributions_subclass
git diff --check && git diff --cached --check
lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I added the `release notes: dynamo` label and reran the five unclassified CI failures. They were all SIGKILLs in `test_serialization_2gb_file`, outside the Dynamo files touched by this PR; I will revisit after the rerun finishes.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR on current main, and tightened the new validation tests to assert the expected distribution validation error. The previous red jobs were repeated SIGKILLs in test_serialization_2gb_file on EC2 linux.2xlarge shadow jobs, unrelated to the Dynamo changes; fresh CI is now running with the current CI routing.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR to pick up the current docs-preview workflow. The previous docs jobs completed docs generation and then failed in the preview upload step because the OSDC job could not find the aws CLI; the newer workflow moves preview upload out of that failing path, and fresh CI is now queued.

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -2739,8 +2803,6 @@ def patch() -> None:
             torch.fx._symbolic_trace.Tracer.trace,
             reason="tracing into FX not fully supported",
         )
-        torch.distributions.Distribution.set_default_validate_args(False)
-
         from torch.optim import (
```
Comment (bobrenjc93) [inline torch/_dynamo/eval_frame.py:2858]: Removing the process-global `set_default_validate_args(False)` here leaves `torch.export(strict=True)` uncovered. With the default `use_new_tracer_experimental=True` (torch/_export/config.py:19), strict export routes through `functional_export._dynamo_graph_capture_for_export`/`dynamo_graph_capture_for_export`, which call `convert_frame.fullgraph_capture` directly (functional_export.py:884, 965) with no `_disable_distribution_validation` scope. Previously this worked only because any prior `torch.compile`/`optimize` in the process had permanently disabled validation via this line. Now, exporting a model that constructs a `torch.distributions.Distribution` (validation on by default, `_validate_args = __debug__`) will enter `constraints.is_dependent`, which is a `SkipFunctionVariable` (trace_rules.py:256) and graph-breaks under fullgraph -> export fails. Please wrap the `fullgraph_capture` calls in functional_export.py with `_disable_distribution_validation()` the same way aot_compile.py:345 does, and add a strict-export-of-distributions regression test (none currently exists under test/export).

Comment (jansel) [reply torch/_dynamo/eval_frame.py:2858]: My agent says thanks for catching the strict export path. I wrapped both functional_export fullgraph_capture call sites with the same _disable_distribution_validation() scope used by AOT compile, and added a strict export distribution regression that covers the default and dynamic-shape capture paths.

**Comment by jansel** [comment]
My agent says I reran the failed TorchTitan features integration job. The failed attempt never reached test execution: the runner pod stayed Pending until the 7200s scheduling timeout, then the container hook errored while cleaning up. I will revisit after the rerun finishes.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
